clients/drcachesim/tools/common/decode_cache.h

```diff
@@ -0,0 +1,469 @@
+        }
+
+        app_pc decode_pc;
+        if (!use_module_mapper_) {
+            decode_pc = const_cast<app_pc>(memref_instr.encoding);
+        } else {
+            // Legacy trace support where we need the binaries.
+            std::string err = find_mapped_trace_address(trace_pc, decode_pc);
+            if (err != "")
+                return err;
+        }
+        app_pc next_pc = decode_from_copy(dcontext_, decode_pc, trace_pc, instr);
+        if (next_pc == nullptr || !instr_valid(instr)) {
+            if (persist_decoded_instrs_) {
+                instr_destroy(dcontext_, instr);
+            }
+            return "decode_from_copy failed";
+        }
+        // Also sets is_valid_.
+        info->second.set_decode_info(dcontext_, memref_instr, instr);
+        return "";
+    }
+
+    /**
+     * Performs initialization tasks such as verifying whether the given trace
+     * indeed has embedded encodings or not, and initializing the
+     * #module_mapper_t if the module path is provided.
+     *
+     * It is important to note that the trace filetype may be obtained using the
+     * get_filetype() API on a #memtrace_stream_t. However, instances of
+     * #memtrace_stream_t have the filetype available at init time (before the
+     * #TRACE_MARKER_TYPE_FILETYPE is actually returned by the stream) only
+     * for offline analysis. In the online analysis case, the user would need to
+     * call this API after init time when the #TRACE_MARKER_TYPE_FILETYPE marker
+     * is seen (which is fine as it occurs before any instr record).
+     *
+     * If the \p module_file_path parameter is not empty, it instructs the
+     * #decode_cache_t object that it should look for the instr encodings in the
+     * application binaries using a #module_mapper_t. A #module_mapper_t is
+     * instantiated only one time and reused for all objects
+     * of #decode_cache_t (of any template type). The user must provide a valid
+     * \p module_file_path if decoding instructions from a trace that does not
+     * have embedded instruction encodings in it, as indicated by absence of
+     * the #OFFLINE_FILE_TYPE_ENCODINGS bit in the #TRACE_MARKER_TYPE_FILETYPE
+     * marker. The user may provide a \p module_file_path also if they
+     * deliberately need to use the module mapper instead of the embedded
+     * encodings.
+     *
+     * If the provided \p module_file_path is empty, the cache object uses
+     * the encodings embedded in the trace records.
+     */
+    std::string
+    init(offline_file_type_t filetype, const std::string &module_file_path = "",
+         const std::string &alt_module_dir = "")
+    {
+        // If we are dealing with a regdeps trace, we need to set the dcontext
+        // ISA mode to the correct synthetic ISA (i.e., DR_ISA_REGDEPS).
+        if (TESTANY(OFFLINE_FILE_TYPE_ARCH_REGDEPS, filetype)) {
+            // Because isa_mode in dcontext is a global resource, we guard its access to
```

Comment:
It's not global, it's limited to that dcontext: but our offline analysis dcontext is global.

Reply:
This was copied as-is from clients/drcachesim/tools/opcode_mix.cpp.

Reworded:

```
            // Because the dcontext used in analysis tools is a shared global resource,
            // we guard its access to avoid data races. Though note that writing to the
            // isa_mode is a benign data race, as all threads are writing the same
            // isa_mode value.
```